src/udp_socket.cpp

```diff
@@ -942,6 +956,58 @@ void socks5::connect2(error_code const& e)
 		TORRENT_ASSERT_FAIL();
 		return;
 	}
+}
+
+void socks5::read_bindaddr(error_code const& e)
+{
+	COMPLETE_ASYNC("socks5::read_bindaddr");
+
+	if (m_abort) return;
+	if (e)
+	{
+		if (m_alerts.should_post<socks5_alert>())
+			m_alerts.emplace_alert<socks5_alert>(m_proxy_addr, operation_t::handshake, e);
+		++m_failures;
+		retry_connection();
+		return;
+	}
+
+	using namespace libtorrent::aux;
+
+	char* p = m_tmp_buf.data();
+	m_udp_proxy_addr.address(address_v4(read_uint32(p)));
+	m_udp_proxy_addr.port(read_uint16(p));
+
+	// we're done!
+	m_active = true;
+	m_failures = 0;
+
+	ADD_OUTSTANDING_ASYNC("socks5::hung_up");
+	boost::asio::async_read(m_socks5_sock, boost::asio::buffer(m_tmp_buf.data(), 10)
+		, std::bind(&socks5::hung_up, self(), _1));
+}
+
+void socks5::read_domainname(error_code const& e)
+{
+	COMPLETE_ASYNC("socks5::read_domainname");
+
+	if (m_abort) return;
+	if (e)
+	{
+		if (m_alerts.should_post<socks5_alert>())
+			m_alerts.emplace_alert<socks5_alert>(m_proxy_addr, operation_t::handshake, e);
+		++m_failures;
+		retry_connection();
+		return;
+	}
+
+	using namespace libtorrent::aux;
+
+	char* p = m_tmp_buf.data();
+	int const len = read_uint8(p);
```

Comment:
No, at this point `p` points at the first hostname character, not length byte. In our case this character is `\0` ...

Reply:
actually, it points to uninitialized memory (because of the bug I mentioned earlier). Note that the `async_read()` calls start at position 1 in the read buffer.